Cheetah/legacy_parser.py

```diff
@@ -469,67 +443,50 @@ def getCallArgString(self):
-        ]
+        For example:
 
-        This method understands *arg, and **kw
+            def foo(bar, baz, x={'womp'}):
+                   ^ (parser is about to parse this)
+
+        Will return "(bar, baz, x={'womp'})" and the parser state afterwards:
+
+            def foo(bar, baz, x={'womp'}):
+                                         ^ (parser is about to parse this)
         """
-        assert self.peek() == '('
-        self.advance()
-        argList = ArgList()
-        onDefVal = False
+        token = self.getPyToken()
+        assert token == '('
+        token_stack = [token]
+        contents = token
 
-        while True:
+        while token_stack:
             if self.atEnd():
                 raise ParseError(
-                    self, msg="EOF was reached before a matching ')'" +
-                    " was found for the '('")
+                    self,
+                    "EOF while searching for '{}' (to match '{}')".format(
+                        closurePairsRev[token_stack[-1]],
+                        token_stack[-1],
+                    )
+                )
 
-            c = self.peek()
-            if c == ')':
-                break
-            elif c in " \t\r\n":
-                if onDefVal:
-                    argList.add_default(c)
-                self.advance()
-            elif c == '=':
-                onDefVal = True
-                self.advance()
-            elif c == ",":
-                argList.next()
-                onDefVal = False
-                self.advance()
-            elif self.startswith(VAR_START):
-                raise ParseError(self, '$ is not allowed here.')
-            elif self.matchIdentifier() and not onDefVal:
-                argList.add_argument(self.getIdentifier())
-            elif onDefVal and c in ('{', '(', '['):
-                argList.add_default(self.getExpression(enclosed=True))
-            elif onDefVal:
-                argList.add_default(self.getPyToken())
-            elif c == '*' and not onDefVal:
-                varName = self.getc()
-                if self.peek() == '*':
-                    varName += self.getc()
-                if not self.matchIdentifier():
-                    raise ParseError(self, 'Expected an identifier.')
-                varName += self.getIdentifier()
-                argList.add_argument(varName)
-            else:
-                raise ParseError(self, 'Unexpected character.')
+            token = self.getPyToken()
+            if token in '({[':
+                token_stack.append(token)
+            elif token in ')}]':
+                if closurePairsRev[token_stack[-1]] != token:
+                    raise AssertionError(
+                        'Mismatched token. '
+                        'Found {} while searching for {}'.format(
+                            token, closurePairsRev[token_stack[-1]],
+                        )
+                    )
+                token_stack.pop()
```

Comment:
We need to figure out what chunk of text to hand to ast.  Or I'm misunderstanding what you're asking.  `getPyToken` uses `tokenize` so this is almost what the ast would do.